File: python/sparklabs/batch/sales/dataframe/orders_revenue.py
from pyspark.sql import functions as F

from sparklabs.batch.sales.dataframe import utils


def run() -> None:
    (
        utils.order_items.join(utils.products, ["product_id"], "inner")
        .withColumn("revenue", F.col("price") * F.col("quantity"))
        .groupBy(F.col("order_id"))
        .agg(
            F.sum("revenue").alias("total_revenue"),
            F.count("*").alias("total_orders"),
        )
        .orderBy(F.desc("total_revenue"))
        .show(truncate=False)
    )


if __name__ == "__main__":
    run()

File: r/sparklabs/batch/sales/dataframe/orders_revenue.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

revenue_df <- withColumn(
  join(order_items, products, order_items$product_id == products$product_id),
  "revenue",
  products$price * order_items$quantity
)
result <- orderBy(
  agg(groupBy(revenue_df, revenue_df$order_id), total_revenue = sum(revenue_df$revenue), total_orders = count(revenue_df$order_id)),
  desc("total_revenue")
)
showDF(result, truncate = FALSE)
